### ISSUE TITLE
Dataset.list_files is impractical for large number of files

### ISSUE THREAD

**Issue by darrengarvey** [tensorflow/tensorflow#17810]
I would like to use [`Dataset.list_files`](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#list_files) on a large dataset. The dataset is similar to ImageNet in that it's a big list of images nested in subdirectories.

I originally guessed that under the hood the api would use queues to start training early, but that isn't the case - the implementation walks the entire directory tree and loads the filenames into memory before the next operation starts.

This means you can wait tens of minutes before training starts.

Two things would help here:
1) Use queues under the hood so downstream ops can start immediately
2) Allow specifying a limit, eg. `Dataset.list_files("**/*.jpg", limit=1000)`

The latter is mostly useful for quicker iteration. The former needs more work due to the different backends for `GetMatchingPaths`.

Is this something being worked on already?

I am using [`os.scandir()`](https://www.python.org/dev/peps/pep-0471/) as a workaround but the `Dataset.list_files()` is a more natural api for this task and should be faster due to not needing to pass the file names via `feed_dict`.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by darrengarvey** [comment]
I'm running TF commit 90df8ab4ab127fac239597ce0708577a335613bf (master as of yesterday). On Ubuntu 17.10. Built with bazel 0.11.1, gcc 6.4, nvcc 9.1 and cuda 9.1. Relevant bazel build options are `-c opt --config=cuda` (ie. -march=native).

However the behaviour I'm talking about is just part of the [`Dataset.list_files()` kernel](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/matching_files_op.cc#L46).

I hacked together a [little script](https://gist.github.com/darrengarvey/ff05fbe28ab2061c101fe64353b467ff) to demonstrate the behaviour. Some results on different disks I have here. All run with `--width 100000`:

1) tmpfs:
```
load data: time: 4566.47 ms
prep data: time: 14.50 ms
read first filename: time: 14691.77 ms
read second filename: time: 0.36 ms
read 99998 more filenames: time: 7742.66 ms (0.08 ms per iteration)
```
2) SSD (Samsung SM951 256GB SSD):
```
load data: time: 9190.51 ms
prep data: time: 19.00 ms
read first filename: time: 20781.54 ms
read second filename: time: 0.28 ms
read 99998 more filenames: time: 7919.55 ms (0.08 ms per iteration)
```

3) External HDD (Samsung D3 Station, 4TB):
```
load data: time: 64532.58 ms
prep data: time: 18.13 ms
read first filename: time: 86195.52 ms
read second filename: time: 0.36 ms
read 99998 more filenames: time: 7742.11 ms (0.08 ms per iteration)
```

4) External HDD (Seagate Expansion Desk 5TB):
```
load data: time: 119427.41 ms
prep data: time: 14.98 ms
read first filename: time: 77205.09 ms
read second filename: time: 0.33 ms
read 99998 more filenames: time: 7723.63 ms (0.08 ms per iteration)
```

Note I'm not flushing caches at any point. Ok so the external drives are terrible, but as a realistic example, consider 1m files sitting on tmpfs:

```
load data: time: 48649.91 ms
prep data: time: 14.79 ms
read first filename: time: 315815.76 ms
read second filename: time: 0.43 ms
read 999998 more filenames: time: 86427.00 ms (0.09 ms per iteration)
```
That's still 5 minutes waiting before training starts.

I suppose I've got a feature request rather than a bug. It might be something I'll fix if no-one else is currently looking at it and the options suggested in the OP seem reasonable.

**Comment by rohan100jain** [comment]
This would be a great feature to have! You're right - currently we wait till the MatchingFilesOp succeeds and we get all the files. One potential idea would be to create an iterator over the output of GetMatchingFiles and stream that into the downstream ops. @mrry for more ideas.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by feihugis** [comment]
@mrry Do you mean to add a new op `MatchingFilesDataset`, which yields results in sorted order?

**Comment by mrry** [comment]
Yes, that's right.

**Comment by feihugis** [comment]
If there is no anyone else working on it, I would like to submit a PR for this new op.

**Comment by mrry** [comment]
Sure, that'd be welcome!

**Comment by feihugis** [comment]
@mrry and @rohan100jain I implemented a `MatchingFilesDatasetOp`, which could yield the matching files in a sorted order.

The performance experiments run using [the script](https://gist.github.com/darrengarvey/ff05fbe28ab2061c101fe64353b467ff) posted by @darrengarvey , where `width` is the number of dirs at each level, and `depth` is the number of nested dirs.

Here are the initial experiment results. What do you think of them?

when `width = 1000` and `depth = 2`:
-- MatchingFilesDataset:
```
read first filename: time: 16.06 ms
read second filename: time: 0.52 ms
read 998 more filenames: time: 336.51 ms (0.34 ms per iteration)
```
-- Dataset.list_files():
```
read first filename: time: 253.91 ms
read second filename: time: 0.27 ms
read 998 more filenames: time: 135.77 ms (0.14 ms per iteration)
```

when `width = 1000` and `depth = 20`:
-- MatchingFilesDataset:
```
read first filename: time: 22.16 ms
read second filename: time: 3.09 ms
read 998 more filenames: time: 2269.14 ms (2.27 ms per iteration)
```

-- Dataset.list_files():
```
read first filename: time: 1860.96 ms
read second filename: time: 0.38 ms
read 998 more filenames: time: 127.23 ms (0.13 ms per iteration)
```

**Comment by darrengarvey** [comment]
Ah, isn't open source fantastic?

@feihugis I hadn't gotten around to implementing this yet so thanks! The results look very promising. So I'm guessing that at each level, you're finding all {files,dirs} in a directory, sorting, then moving to the next level? This sounds great and the results are a big improvement.

Is the sorting optional? In many cases the order isn't interesting so if it could be turned off for extra speed that would be worth a small fortune in virtual bonus points. Sorting by default does seem sensible though.

Kudos!

**Comment by feihugis** [comment]
@darrengarvey Yeah, I implemented it in a similar way. Basically, I use `DFS` to search the files instead of `BFS` used in `MatchingFilesOp`. The files and dirs at each level are stored in a `priority_queue`, then the output for each `get_next` could be sorted. 

Sorting is not optional yet. I could make some change to make it optional. Will do some tests to see the performance difference.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by darrengarvey** [comment]
This is fixed now #22429 is merged. Thanks @feihugis.
